src/parse_execution.rs

```diff
@@ -175,15 +175,15 @@ impl<'a> ParseExecutionContext {
     pub fn eval_node(
         &self,
         ctx: &OperationContext<'_>,
-        node: &dyn Node,
+        node: &impl Node,
         associated_block: Option<BlockId>,
     ) -> EndExecutionReason {
-        match node.typ() {
-            ast::Type::statement => {
-                self.eval_statement(ctx, node.as_statement().unwrap(), associated_block)
+        match node.as_node() {
```

Comment:
This is really nice overall, especially for writing match statements; LSP type definition is useful for `NodeRef` since it jumps to the enum node hierarchy.

It's a bit ugly that we have to use `node.as_node()` to convert from impl Node to the enum_dispatched enum.
Do you think it would be possible to get rid of the traits altogether and pass around something like `NodeRef`?

Combining traits with enum_dispatch means we need all those blanket trait impls for references.
I guess it's no big deal but since enum_dispatch doesn't need traits, I wonder if we still do?

For the methods we need to call generically, can't we call them `NodeEnum` already?
I guess it would be a bit ugly to copy all the `trait List` default impls
to every single list node but it's not like `define_list_node` is small either way.

Can we look at the cases where we inherently need traits, and maybe use enum dispatch everywhere else?
I know that the parent node pointer is one case where we need to use virtual dispatch,
but I don't think that applies to subtraits like Leaf.
I'm not sure what's the answer here but maybe you have enough context to find it.

Reply:
Future-work bullet point 3: This code path will eventually use either a narrower `EvaluatableNodeRef<'a>` enum or a generic bounded by a new `EvaluatableNode` trait (names absolutely not final).

Reply:
There is definitely room to _reduce_ our trait surface area, but the way to make that happen is somewhat complex. There are a few cases where it genuinely makes more sense to use a trait, as each callsite knows the exact concrete type it's passing and would have to upcast.

Reply:
> There are a few cases where it genuinely makes more sense to use a trait, as each callsite knows the exact concrete type it's passing and would have to upcast.

Do you have an example?

Reply:
A lot of stuff in `highlight` is the main thing that comes to mind. For example, `color_as_argument` doesn't do any downcasting and has three callsites, each of which has a known concrete type.

Reply:
I don't think `color_as_argument` qualifies; one reason is that it doesn't read its `node: impl Node` argument except for `source_range()` which we could compute at the call site. The other reason is that we want the `Node` trait regardless (because of parent pointers).
So only the other traits are relevant, we might want to get rid of the `Leaf` trait because we can pass a `LeafEnum` instead.

Reply:
Point taken. My drafts of the `trait Parse` refactor do shrink down some of the traits significantly, which should put us in a better position to experiment with removing traits and/or swapping out `impl Trait` parameters for enums.